### PR TITLE
"empty response" not logged as error in memcache discovery client

### LINKED ISSUE #146: Empty response warning with latest kube-client (v0.26)
[Recent changes](https://github.com/kubernetes/client-go/commit/25d5761bfa0603c77cf5efefb14fd2e63a03a773) in https://github.com/kubernetes/client-go have changed the behavior of discovery client. 
These changes are released as part of v0.26 and produced a scenario where the tooling like kubectl or helm shows a warning during the resource discovery:
```
E1212 15:54:39.565250   21101 memcache.go:255] couldn't get resource list for external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1
```
As a first action, I reviewed if we are doing something wrong in KEDA but after debugging helm and kubectl I found that the reason isn't related with KEDA (prometheus-adapter produces the same error but about custom.metrics) itself.
After the changes, the discovery client list all the api resources and also all the resources inside the api resources, so the scenario where a fresh installation of KEDA/prometheus-adapter without any metric exposed automatically shows this warning on every update tooling, which is annoying (and looks hideous).

In KEDA, [we return the current metrics available](https://github.com/kedacore/keda/blob/main/pkg/provider/provider.go#L212) in response of listing requests, and I have seen that other tools like prometheus-adapter does the same for [external metrics](https://github.com/kubernetes-sigs/prometheus-adapter/blob/master/pkg/external-provider/provider.go#L65) and [custom metrics](https://github.com/kubernetes-sigs/prometheus-adapter/blob/master/pkg/custom-provider/series_registry.go#L134)

Is this approach the correct? I mean, should we return the metrics instead of the metric type (like metrics.k8s.io does). Exposing all the metrics available as part of listing requests can produce timeouts if the amount of available metrics is huge, and can produce this annoying (and unrelated) warning in case of non metrics for scaling available.
I can see that this approach is also used in the [example code](https://github.com/kubernetes-sigs/custom-metrics-apiserver/blob/master/test-adapter/provider/provider.go#L348), but if this is how we should do the things exposing metrics, the discovery client should consider this to ignore those errors.

Could you provide any help/guidance about this topic. We have received some notifications about this warning in KEDA project and users think it's related with a KEDA problem, but we follow the available example to develop our metrics server.

Is correct from custom-metrics-apiserver usage if we return a fixed collection of generic types (in our case, `scaledobjects`) like metrics.k8s.io api does (with 'pods' and 'nodes'). This change would solve the timeouts in huge clusters and also the issue without any metric. Considering that these metrics aren't browsable like metric-k8s-io are, maybe this could be acceptable from custom-metrics-apiserver implementation pov

### LINKED ISSUE #4224: couldn't get resource list for external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1
### Report

Well, it happens all the time, **I can't use KEDA** locally for testing.

The error : 
```log
E0210 10:57:59.544134  457383 memcache.go:255] couldn't get resource list for external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1
```

Tried on:
- Inside a VM
  - minikube
    - metrics-server as addon
    - metrics-server from helm
    - keda from helm
    - keda from apply
  - docker desktop
    - metrics-server from helm
    - keda from helm
    - keda from apply
  - microk8s
    - metrics-server as addon
    - metrics-server from helm
    - keda from helm
    - keda from apply
- Bare metal
  - docker desktop
    - metrics-server from helm
    - keda from helm
    - keda from apply

### Expected Behavior

Proper communication with metric-server and presenting the resources.

### Actual Behavior

```sh
kubectl get --raw "/apis/external.metrics.k8s.io/v1beta1"
{"kind":"APIResourceList","apiVersion":"v1","groupVersion":"external.metrics.k8s.io/v1beta1","resources":[]}
```

I've made some stress tests and the deployments weren't being scaled.

### Steps to Reproduce the Problem

Simple example :
```sh
# start any kubectl context : minikube, docker-desktop, microk8s

# enable metrics-server
# or simply run this
helm upgrade --install metrics-server metrics-server/metrics-server -n kube-system --set args={--kubelet-insecure-tls}

# wait and make sure that HPA and "top nodes" are working

# deploy keda
# add its repo from website... then
kubectl create namespace keda
helm install keda kedacore/keda -n keda
```

After that, most of the times simply running `kubectl get all -n keda` is enough to raise the error. In a very few attempts I had to try creating a _ScaledObject_ and then it happened immediately. I can't even check if it is a problem with my _ScaledObject_ because the error keeps occurring even after deleting the objects.

### Logs from KEDA operator

```sh
microk8s kubectl get pods -n keda
E0210 17:12:43.288248 1396573 memcache.go:255] couldn't get resource list for external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1
NAME                                      READY   STATUS    RESTARTS   AGE
keda-metrics-apiserver-85f98c6655-bq2zn   1/1     Running   0          21m
keda-operator-97b74b8c8-xjchw             1/1     Running   0          21m


microk8s kubectl logs -n keda keda-operator-97b74b8c8-xjchw -c keda-operator
E0210 17:12:45.963719 1396717 memcache.go:255] couldn't get resource list for external.metrics.k8s.io/v1beta1: Got empty response for: external.metrics.k8s.io/v1beta1
2023-02-10T19:51:26Z    INFO    controller-runtime.metrics      Metrics server is starting to listen    {"addr": ":8080"}
2023-02-10T19:51:26Z    INFO    setup   Starting manager
2023-02-10T19:51:26Z    INFO    setup   KEDA Version: 2.9.2
2023-02-10T19:51:26Z    INFO    setup   Git Commit: 9bc3f66578a08cdfe084468ea3ef998fa6bf3bb0
2023-02-10T19:51:26Z    INFO    setup   Go Version: go1.18.8
2023-02-10T19:51:26Z    INFO    setup   Go OS/Arch: linux/amd64
2023-02-10T19:51:26Z    INFO    setup   Running on Kubernetes 1.26      {"version": "v1.26.1"}
I0210 19:51:26.316433       1 leaderelection.go:248] attempting to acquire leader lease keda/operator.keda.sh...
2023-02-10T19:51:26Z    INFO    Starting server {"path": "/metrics", "kind": "metrics", "addr": "[::]:8080"}
2023-02-10T19:51:26Z    INFO    Starting server {"kind": "health probe", "addr": "[::]:8081"}
I0210 19:51:26.354840       1 leaderelection.go:258] successfully acquired lease keda/operator.keda.sh
2023-02-10T19:51:26Z    INFO    Starting EventSource    {"controller": "scaledobject", "controllerGroup": "keda.sh", "controllerKind": "ScaledObject", "source": "kind source: *v1alpha1.ScaledObject"}
2023-02-10T19:51:26Z    INFO    Starting EventSource    {"controller": "scaledobject", "controllerGroup": "keda.sh", "controllerKind": "ScaledObject", "source": "kind source: *v2.HorizontalPodAutoscaler"}
2023-02-10T19:51:26Z    INFO    Starting Controller     {"controller": "scaledobject", "controllerGroup": "keda.sh", "controllerKind": "ScaledObject"}
2023-02-10T19:51:26Z    INFO    grpc_server     Starting Metrics Service gRPC Server    {"address": ":9666"}
2023-02-10T19:51:26Z    INFO    Starting EventSource    {"controller": "triggerauthentication", "controllerGroup": "keda.sh", "controllerKind": "TriggerAuthentication", "source": "kind source: *v1alpha1.TriggerAuthentication"}
2023-02-10T19:51:26Z    INFO    Starting Controller     {"controller": "triggerauthentication", "controllerGroup": "keda.sh", "controllerKind": "TriggerAuthentication"}
2023-02-10T19:51:26Z    INFO    Starting EventSource    {"controller": "clustertriggerauthentication", "controllerGroup": "keda.sh", "controllerKind": "ClusterTriggerAuthentication", "source": "kind source: *v1alpha1.ClusterTriggerAuthentication"}
2023-02-10T19:51:26Z    INFO    Starting Controller     {"controller": "clustertriggerauthentication", "controllerGroup": "keda.sh", "controllerKind": "ClusterTriggerAuthentication"}
2023-02-10T19:51:26Z    INFO    Starting EventSource    {"controller": "scaledjob", "controllerGroup": "keda.sh", "controllerKind": "ScaledJob", "source": "kind source: *v1alpha1.ScaledJob"}
2023-02-10T19:51:26Z    INFO    Starting Controller     {"controller": "scaledjob", "controllerGroup": "keda.sh", "controllerKind": "ScaledJob"}
2023-02-10T19:51:26Z    INFO    Starting workers        {"controller": "clustertriggerauthentication", "controllerGroup": "keda.sh", "controllerKind": "ClusterTriggerAuthentication", "worker count": 1}
2023-02-10T19:51:26Z    INFO    Starting workers        {"controller": "triggerauthentication", "controllerGroup": "keda.sh", "controllerKind": "TriggerAuthentication", "worker count": 1}
2023-02-10T19:51:26Z    INFO    Starting workers        {"controller": "scaledjob", "controllerGroup": "keda.sh", "controllerKind": "ScaledJob", "worker count": 1}
2023-02-10T19:51:26Z    INFO    Starting workers        {"controller": "scaledobject", "controllerGroup": "keda.sh", "controllerKind": "ScaledObject", "worker count": 5}
```

### KEDA Version

2.9.2

### Kubernetes Version

1.26

### Platform

None

### Scaler Details

CPU, Memory

### Anything else?

Yes, I have _**ALREADY**_ seen the [workaround comment on another issue](https://github.com/kubeshop/botkube/issues/829#issuecomment-1302215298) telling to "create a dummy resource".

I could be missing something, but as far as I can tell, the resources should be presented to me and if it is mandatory to do something else, it should be **very explicit** in the docs.

I don't have any idea of what I should to to create this dummy resource and what exactly that means.

### LINKED ISSUE #111292: Discovery memCacheClient has inconsistent error behavior with other DiscoveryInterfaces
### What happened?

I think this got stale and closed because it was posted on `kubernetes/client-go` but is still very relevant:
https://github.com/kubernetes/client-go/issues/745

The memCacheClient created by memory.NewMemCacheClient(...) returns an error when calling ServerResourcesForGroupVersion if the APIResources list returned is empty, as seen here:
https://github.com/kubernetes/client-go/blob/6d14209323d355841b67c4d9ca8c22fb1ca930a6/discovery/cached/memory/memcache.go#L214-L216

Our particular case came up due to the `external.metrics.k8s.io/v1beta1` API group being registered with the API server, but having no resources.

### What did you expect to happen?

If using [NewMemcachedDiscoveryClient](https://github.com/kubecfg/kubecfg/blob/4b23b794b8281c7886b5fd0bd03d7a38db876c10/cmd/root.go#L425) like [here](https://github.com/kubecfg/kubecfg/blob/4b23b794b8281c7886b5fd0bd03d7a38db876c10/pkg/kubecfg/validate.go#L48) to discover `apiservices` with zero current resources there should not be an error returned.

### How can we reproduce it (as minimally and precisely as possible)?

Do a discovery with a `NewMemcachedDiscoveryClient` against a Kubernetes cluster that has at least one `apiservice` with zero resources. The condition on the server that triggers the described issue can be expressed as e.g.:
```
$ kubectl get --raw /apis/external.metrics.k8s.io/v1beta1
{"kind": "APIResourceList","apiVersion":"v1","groupVersion":"external.metrics.k8s.io/v1beta1","resources":[]}
```

### Anything else we need to know?

I would be good to get to know the reasoning behind the emptiness check in the `memCacheClient`, maybe there is good reason for it.

### Kubernetes version

<details>
Client Version: v1.20.15
Server Version: v1.22.8-gke.202
</details>

### Cloud provider

<details>
Google Cloud (GKE)
</details>

### OS version

_No response_

### Install tools

_No response_

### Container runtime (CRI) and version (if applicable)

_No response_

### Related plugins (CNI, CSI, ...) and versions (if applicable)

_No response_

### PR DESCRIPTION
**Pull request by seans3**
* For memory cached discovery client, logs discovery "empty response" at `klog.V(3)` (same as disk cache client) instead of as an error. Adds custom error struct `emptyResponseError` for this purpose.
* Fixes issue where empty list of resources returned by poorly behaved api servers creates excessive error logging. In particular, metrics aggregated api servers regularly return an empty list of resources for GroupVersion discovery requests. Error logging has significantly increased since the aggregated discovery change in `v1.26.0` released in Dec. 2022: https://github.com/kubernetes/kubernetes/pull/113599

/kind bug

Fixes https://github.com/kubernetes/kubernetes/issues/111292
Fixes https://github.com/kubernetes-sigs/custom-metrics-apiserver/issues/146
Fixes https://github.com/kedacore/keda/issues/4224

```release-note
Fixes a 1.26 regression in client-go discovery handling of empty responses
```

### CODE REVIEW

**Comment by liggitt** [comment]
> * Does _not_ change behavior--only changes error logging.

this _does_ change behavior, it means we cache and return zero-length non-error entry and don't return an error from ServerResourcesForGroupVersion any more

**Comment by liggitt** [comment]
making memory-cache and disk-cache consistent seems like a good idea, but since our controllers almost exclusively use the memory-based cache, we still need to think through what impact this changed behavior will have on them

can we look at why this check was put in place initially, and make sure we're not opening ourselves up to failure modes that were previously fixed?

**Comment by JorTurFer** [comment]
Could we just exclude `custom.metrics.k8s.io` and `external.metrics.k8s.io` from the check? I guess that those are the aggregated apis which can dynamically change based on user actions where 0 length is a correct value (because users haven't registered any metric yet)

**Comment by liggitt** [comment]
> Could we just exclude `custom.metrics.k8s.io` and `external.metrics.k8s.io` from the check?

no, we wouldn't hard-code specific aggregated APIs here

File: staging/src/k8s.io/client-go/discovery/cached/memory/memcache.go [hunk 1]
```diff
@@ -103,7 +112,10 @@ func (d *memCacheClient) ServerResourcesForGroupVersion(groupVersion string) (*m
 	if cachedVal.err != nil && isTransientError(cachedVal.err) {
 		r, err := d.serverResourcesForGroupVersion(groupVersion)
 		if err != nil {
-			utilruntime.HandleError(fmt.Errorf("couldn't get resource list for %v: %v", groupVersion, err))
+			// Do not log errors for "empty response", which is a common response for metrics.
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/discovery/cached/memory/memcache.go:115]: should we log empty errors at V(3) verbosity like diskcache?

File: staging/src/k8s.io/client-go/discovery/cached/memory/memcache.go [hunk 2]
```diff
@@ -252,7 +264,10 @@ func (d *memCacheClient) refreshLocked() error {
 
 				r, err := d.serverResourcesForGroupVersion(gv)
 				if err != nil {
-					utilruntime.HandleError(fmt.Errorf("couldn't get resource list for %v: %v", gv, err))
+					// Do not log errors for "empty response", which is a common response for metrics.
+					if _, empty := err.(*emptyResponseError); !empty {
```
Comment (liggitt) [inline staging/src/k8s.io/client-go/discovery/cached/memory/memcache.go:268]: should we log empty errors at V(3) verbosity like diskcache?

File: staging/src/k8s.io/client-go/discovery/cached/memory/memcache.go (hunk 1 above)
Comment (seans3) [reply staging/src/k8s.io/client-go/discovery/cached/memory/memcache.go]: Done

**Comment by zroubalik** [comment]
Thanks for merging this PR, is ti going to be backported to 1.26 as well?

**Comment by seans3** [comment]
> Jordan Liggitt Thanks for merging this PR, is it going to be backported to 1.26 as well?

https://github.com/kubernetes/kubernetes/pull/116352

**Comment by rodrigc** [comment]
@seans3 I'm having problems with keda, which I _think_ might be solved by this PR.

I asked this question about versions: https://github.com/kedacore/keda/issues/4224#issuecomment-1524001368

I don't know what version of kubectl and the kubernetes server I need to fully have this fix for keda.

@sean3 if you have an idea, can you please shed some light?

**Pull request merged**